File: ios/Sources/EdgeFrameworks/Runtime/EdgeAgent.swift
import Foundation

public enum EdgeAgentError: Error, Equatable {
    case noCompatibleProvider
}

public struct EdgeAgent: Sendable {
    private let router: EdgeProviderRouter

    public init(router: EdgeProviderRouter) {
        self.router = router
    }

    public func run(
        _ request: EdgeGenerationRequest,
        requiring capabilities: Set<EdgeCapability> = [.textGeneration]
    ) async throws -> EdgeGenerationResponse {
        guard let provider = await router.provider(supporting: capabilities) else {
            throw EdgeAgentError.noCompatibleProvider
        }

        return try await provider.generate(request)
    }

    public func stream(
        _ request: EdgeGenerationRequest,
        requiring capabilities: Set<EdgeCapability> = [.textGeneration, .streaming]
    ) async throws -> AsyncThrowingStream<EdgeGenerationEvent, Error> {
        guard let provider = await router.provider(supporting: capabilities) else {
            throw EdgeAgentError.noCompatibleProvider
        }

        return provider.stream(request)
    }
}

File: android/edge-frameworks-core/src/main/kotlin/io/github/rajibsarwar/edgeframeworks/EdgeAgent.kt
package io.github.rajibsarwar.edgeframeworks

import kotlinx.coroutines.flow.Flow

class EdgeAgent(
    private val router: EdgeProviderRouter
) {
    suspend fun run(
        request: EdgeGenerationRequest,
        requiredCapabilities: Set<EdgeCapability> = setOf(
            EdgeCapability.TEXT_GENERATION
        )
    ): EdgeGenerationResponse {
        val provider = router.provider(requiredCapabilities)
            ?: throw EdgeAgentException.NoCompatibleProvider

        return provider.generate(request)
    }

    suspend fun stream(
        request: EdgeGenerationRequest,
        requiredCapabilities: Set<EdgeCapability> = setOf(
            EdgeCapability.TEXT_GENERATION,
            EdgeCapability.STREAMING
        )
    ): Flow<EdgeGenerationEvent> {
        val provider = router.provider(requiredCapabilities)
            ?: throw EdgeAgentException.NoCompatibleProvider

        return provider.stream(request)
    }
}

sealed class EdgeAgentException(message: String) : Exception(message) {
    data object NoCompatibleProvider :
        EdgeAgentException("No compatible model provider is registered")
}
